gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2231,6 +2258,396 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 	return nn
 }
 
+// Identifies NameExprTypeLoopDefines.
+// Also finds GotoLoopStmts, XXX but probably remove, not needed.
+func findGotoLoopDefines(ctx BlockNode, bn BlockNode) {
+	// create stack of BlockNodes.
+	var stack []BlockNode = make([]BlockNode, 0, 32)
+	var last BlockNode = ctx
+	stack = append(stack, last)
+
+	// iterate over all nodes recursively.
+	_ = Transcribe(bn, func(ns []Node, ftype TransField, index int, n Node, stage TransStage) (Node, TransCtrl) {
+		defer func() {
+			if r := recover(); r != nil {
+				// before re-throwing the error, append location information to message.
+				loc := last.GetLocation()
+				if nline := n.GetLine(); nline > 0 {
+					loc.Line = nline
+				}
+
+				var err error
+				rerr, ok := r.(error)
+				if ok {
+					// NOTE: gotuna/gorilla expects error exceptions.
+					err = errors.Wrap(rerr, loc.String())
+				} else {
+					// NOTE: gotuna/gorilla expects error exceptions.
+					err = fmt.Errorf("%s: %v", loc.String(), r)
+				}
+
+				// Re-throw the error after wrapping it with the preprocessing stack information.
+				panic(&PreprocessError{
+					err:   err,
+					stack: stack,
+				})
+			}
+		}()
+
+		if debug {
+			debug.Printf("findGotoLoopDefines %s (%v) stage:%v\n", n.String(), reflect.TypeOf(n), stage)
+		}
+
+		switch stage {
+		// ----------------------------------------
+		case TRANS_ENTER:
+			return n, TRANS_CONTINUE
+
+			// ----------------------------------------
+		case TRANS_BLOCK:
+			if bn, ok := n.(BlockNode); ok {
+				pushInitBlock(bn, &last, &stack)
+			}
+
+		// ----------------------------------------
+		case TRANS_LEAVE:
+			switch n := n.(type) {
+			case *BranchStmt:
+				switch n.Op {
+				case GOTO:
+					// bn, depth, index, labelLine := findGotoLabel(last, n.Label)
+					bn, _, _, labelLine := findGotoLabel(last, n.Label)
+					// already done in Preprocess:
+					// n.Depth = depth
+					// n.BodyIndex = index
+
+					// this goto line location
+					gotoLine := n.GetLine()
+					// skip if destination comes after.
+					if labelLine >= gotoLine {
+						return n, TRANS_SKIP
+					}
+
+					// recurse and mark stmts as ATTR_GOTOLOOP_STMT
+					// We don't really need ATTR_GOTOLOOP_STMT,
+					// but the logic is here and might use it one day,
+					// so keep for a while.
+					Transcribe(bn,
+						func(ns []Node, ftype TransField, index int, n Node, stage TransStage) (Node, TransCtrl) {
+							switch stage {
+							case TRANS_ENTER:
+								// NOTE: called redundantly
+								// for many goto stmts,
+								// idempotenct updates only.
+								switch n := n.(type) {
+								// Skip the body of these:
+								case *FuncLitExpr:
+									return n, TRANS_SKIP
+								case *FuncDecl:
+									return n, TRANS_SKIP
```

Comment:
in this case, the func `Foo` will be skipped. 
```go
func Foo(){
...
LABEL:
  xxx
  goto LABEL
```

Reply:
yes.